KAFKA-5135: Controller Health Metrics (KIP-143)

ijuma opened the pull request.

ijuma committed:
Add timer metrics

ijuma committed:
Code clean-ups

ijuma committed:
Add controller state metric

ijuma committed:
Use volatile variables instead of `AtomicInteger` and other clean-ups

We don’t use any functionality from `AtomicInteger`, we simply set
and read values.

ijuma committed:
Add ControllerChannelManager queue size metrics

ijuma committed:
Add FailedIsrUpdatesPerSec metric

ijuma commented:
This is an initial PR to get some feedback on whether I'm on the right track. Need to write tests and verify if any additional work needs to be done to clean-up metrics. cc @onurkaraman @junrao

onurkaraman commented on core/src/main/scala/kafka/controller/ControllerChannelManager.scala:
```diff
@@ -38,11 +40,24 @@ import scala.collection.JavaConverters._
+  newGauge(
+    "TotalQueueSize",
+    new Gauge[Int] {
+      def value: Int = brokerStateInfo.values.iterator.map(_.messageQueue.size).sum
```
This is a concurrency bug. You should be synchronizing access to brokerStateInfo.

onurkaraman commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1713,16 +1747,31 @@ case class LeaderIsrAndControllerEpoch(leaderAndIsr: LeaderAndIsr, controllerEpo
+  case object Idle extends ControllerState(0)
+  case object Starting extends ControllerState(1)
+  case object Resigning extends ControllerState(2)
+  case object ElectingLeader extends ControllerState(3)
```
The `ElectingLeader` state seems somewhat out of place. Should this reflect that a broker change event is occurring?

ijuma replied:
The KIP says `BrokerChange` indeed. However, we have `LeaderElectionRateAndTimeMs` that seems to measure the same code path. So, I thought it would make sense to use a similar name.

onurkaraman commented on core/src/main/scala/kafka/server/KafkaRequestHandler.scala:
```diff
@@ -132,7 +132,7 @@ class BrokerTopicMetrics(name: Option[String]) extends KafkaMetricsGroup {
 }
 
-object BrokerTopicStats extends Logging {
+object BrokerTopicStats {
```
The `BrokerTopicStats` refactoring causes the PR to be much larger than I think it needs to be. Can we move this change out into a separate cleanup PR?

ijuma replied:
The changes are mechanical,  the diffstat is +483 −315 (which is manageable) and there's a separate commit for the refactoring. So, I think we don't need a separate PR. Let's see what @junrao says.

ijuma committed:
Hold brokerLock while accessing brokerStateInfo

junrao commented:
@ijuma : Thanks for the patch. The general approach looks good.

ijuma committed:
Include KIP updates and use a hopefully better approach.

junrao commented on core/src/main/scala/kafka/controller/KafkaController.scala:
```diff
@@ -1097,39 +1100,37 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
-    debug("preferred replicas by broker " + preferredReplicasForTopicsByBrokers)
+  private def checkAndTriggerAutoLeaderRebalance(): Unit = {
+    trace("Checking need to trigger auto leader balancing")
+    // get all the active brokers
```
The comment seems inaccurate.

ijuma replied:
No idea how that ended there, maybe a copy and paste malfunction. Great catch.

ijuma committed:
Extract ControllerEventManager for easier testing

ijuma committed:
Make `KafkaController.eventManager` private and pass it to classes that need it

Reducing exposure of fields from KafkaController reduces coupling.

ijuma committed:
Add unit test and modify an existing integration test

ijuma commented on core/src/main/scala/kafka/controller/ControllerEventManager.scala:
```diff
@@ -0,0 +1,64 @@
+        case e: Throwable => error(s"Error processing event $controllerEvent", e)
+      }
+
+      eventProcessedListener(controllerEvent)
```
We are calling this outside the try/catch as it was before this PR. One potential issue is that if an exception is thrown by this method, the thread will die.

In a sense this is good because it helps us catch issues during testing (unlike `process()`, this code should never throw exceptions). The downside is that if we fail to catch such issues during testing, it will take the whole Controller down. Given that, would it be better to enclose this in a try/catch?

junrao replied:
Yes, it does seem that including it in try/catch is better.

ijuma commented on core/src/main/scala/kafka/controller/ControllerState.scala:
```diff
@@ -0,0 +1,80 @@
+  protected def hasRateAndTimeMetric: Boolean = true
+}
+
+object ControllerState {
```
The names below a little inconsistent although they match the KIP. More concretely: Idle, ControllerChange, BrokerChange, TopicChange, IsrChange, ControlledShutdown, TopicDeletion versus PartitionReassigning, AutoLeaderBalancing, ManualLeaderBalancing. It seems like we should do either of the following:

1. Change the former to be: Idling, ControllerChanging, BrokerChanging, etc.
2. Change the latter to be: PartitionReassignment, AutoLeaderBalance, ManualLeaderBalance

Thoughts?

junrao replied:
It seems that #2 is better?

junrao commented on core/src/main/scala/kafka/controller/ControllerState.scala:
```diff
@@ -0,0 +1,80 @@
+  }
+
+  case object BrokerChange extends ControllerState {
+    def value = 2
```
Since we have an existing metric ControllerStats.leaderElectionTimer, for this state, perhaps it's better not to measure it again in ControllerEventThread?

ijuma replied:
Yeah, I was unsure about consistency and clarity over duplication. I'll do as you suggest for now, as we can add a `BrokerChange` metric in the future, if this proves to be confusing.

junrao reviewed (commented):
@ijuma : Thanks for the patch. LGTM. Just a couple of minor comments.

onurkaraman commented on core/src/main/scala/kafka/controller/ControllerEventManager.scala:
```diff
@@ -0,0 +1,64 @@
+      _state = controllerEvent.state
+
+      try {
+        rateAndTimeMetrics(state).time {
```
Nice! This is exactly what I was hoping the loop would look like.

ijuma committed:
Rename some controller states so that they are all nouns

ijuma committed:
Don’t duplicate existing `LeaderElectionRateAndTimeMs` metric

ijuma committed:
Remove comment that was added by mistake

ijuma committed:
Catch and log if ControllerEventManager listener throws exception

ijuma committed:
Remove redundant try/catch from some ControllerEvent implementations

ijuma commented:
Thanks for the review @junrao. I addressed your feedback. Also, while removing the `controllerContext.stats.leaderElectionTimer.time` block (which is now handled via the `BrokerChange` state metric), I noticed that some events had a redundant try/catch block (since `ControllerEventManager.process` also does the same). I've removed them as well.

junrao commented on core/src/test/scala/unit/kafka/controller/ControllerEventManagerTest.scala:
```diff
@@ -0,0 +1,94 @@
+    EasyMock.expect(eventMock.process()).andAnswer(new IAnswer[Unit]() {
+      def answer(): Unit = {
+        while (processing)
+          Thread.`yield`()
```
Since the behavior of Thread.yield() can be platform dependent, would it be better to change processing to sth like a CountDownLatch()?

ijuma replied:
That's a good suggestion. Both are correct (even if `Thread.yield` is a no-op, it won't matter here), but `CountDownLatch` is more concise and shows intent better. Updated the PR.

junrao reviewed (commented):
@ijuma : Thanks for the updated patch. LGTM. Just one minor comment.

ijuma committed:
Use `CountDownLatch` to simplify the code

junrao commented:
@ijuma : Thanks for the patch. LGTM. I will let you merge this when the tests complete.
